tools/binary2cpp.cpp

```diff
@@ -13,49 +13,79 @@
 // chars. Similar to the xxd utility.
 
 static int usage() {
-    fprintf(stderr, "Usage: binary2cpp identifier [-header]\n");
+    fprintf(stderr, "Usage: binary2cpp identifier [-header] [-hidden]\n");
     return -1;
 }
 
+void emit_hidden_macro(const char *target) {
+    // Note that visibility=hidden isn't supported under MinGW, regardless of compiler
+    printf("#if defined(_MSC_VER) || defined(__MINGW32__) || defined(__MINGW64__)\n");
+    printf("#define HALIDE_BINARY2CPP_%s_HIDDEN_DATA\n", target);
+    printf("#else\n");
+    printf("#define HALIDE_BINARY2CPP_%s_HIDDEN_DATA __attribute__((visibility(\"hidden\")))\n", target);
+    printf("#endif\n");
+}
+
 int main(int argc, const char **argv) {
-    const char *target = argv[1];
-    if (argc == 3) {
-        if (!strcmp(argv[2], "-header")) {
-            printf("#ifndef _H_%s_binary2cpp\n", target);
-            printf("#define _H_%s_binary2cpp\n", target);
-            printf("extern \"C\" {\n");
-            printf("extern unsigned char %s[];\n", target);
-            printf("extern int %s_length;\n", target);
-            printf("}  // extern \"C\"\n");
-            printf("#endif  // _H_%s_binary2cpp\n", target);
-            return 0;
-        } else {
-            return usage();
-        }
-    } else if (argc > 3) {
+    if (argc < 2 || argc > 4) {
         return usage();
     }
 
-#ifdef _WIN32
-    setmode(fileno(stdin), O_BINARY); // On windows bad things will happen unless we read stdin in binary mode
-#endif
-    printf("extern \"C\" {\n");
-    printf("unsigned char %s[] = {\n", target);
-    int count = 0;
-    int line_break = 0;
-    while (1) {
-        int c = getchar();
-        if (c == EOF) break;
-        printf("0x%02x, ", c);
-        // Not necessary, but makes a bit easier to read
-        if (++line_break > 12) {
-            printf("\n");
-            line_break = 0;
+    const char *target = "";
+    bool header = false;
+    bool hidden = false;
+
+    for (int i = 1; i < argc; ++i) {
+        if (!strcmp(argv[i], "-header")) header = true;
+        else if (!strcmp(argv[i], "-hidden")) hidden = true;
+        else if (argv[i][0] == '-') { return usage(); }
+        else target = argv[i];
+    }
+    if (!target[0]) {
+        return usage();
+    }
+
+    char vis[1024] = "";
+    if (hidden) {
+        sprintf(vis, "HALIDE_BINARY2CPP_%s_HIDDEN_DATA", target);
+    }
+
+    if (header) {
+        printf("#ifndef _H_%s_binary2cpp\n", target);
+        printf("#define _H_%s_binary2cpp\n", target);
+        if (hidden) {
+            emit_hidden_macro(target);
+        }
+        printf("extern \"C\" {\n");
+        printf("extern %s unsigned char %s[];\n", vis, target);
```

Comment:
I don't think you need the visibility attribute in the header, only on the definition (possibly only on one of the declarations regardless of whether it is a definition or not). Does that make it easier to generate this file at all? It seems like it might?

Reply:
Yeah, probably. (Though AFAICT we don't currently ever use the -header mode, so it's probably moot...)

Reply:
Interestingly, from a quick google, I can't find definitive documentation on what happens if you get these attributes mismatched between declaration and definition. Though I guess the answer is likely "it's fine" since we're essentially doing extern redeclarations of these, without the attribute, in LLVM_Runtime_Linker.cpp.